File: kapun-credential-core/src/commonMain/kotlin/org/kapunsdk/ClaimsPointer.kt
/* Copyright 2025 Ubique Innovation AG

Licensed under the Apache License, Version 2.0 (the "License");
you may not use this file except in compliance with the License.
You may obtain a copy of the License at

  http://www.apache.org/licenses/LICENSE-2.0

Unless required by applicable law or agreed to in writing,
software distributed under the License is distributed on an
"AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
KIND, either express or implied.  See the License for the
specific language governing permissions and limitations
under the License.
 */

package org.kapunsdk.credentials

import org.kapunsdk.util.extensions.get
import org.kapunsdk.util.extensions.getAll
import org.kapunsdk.util.extensions.isArrayLike
import org.kapunsdk.util.extensions.isObject
import uniffi.kapun_util_rust.Value
import uniffi.kapun_credential_core_rust.PointerPart
import uniffi.kapun_credential_core_rust.QueryException
import uniffi.kapun_credential_core_rust.Selector

class ClaimsPointer(val path: List<PointerPart>) : Selector {
    fun fromDepth(depth: Int) : ClaimsPointer {
        return ClaimsPointer(path.slice(depth-1..<path.size))
    }
    fun toDepth(depth: Int) : ClaimsPointer {
        return ClaimsPointer(path.slice(0..<depth))
    }
    fun key() : PointerPart {
        return path.last()
    }
    fun depth() : Int {
        return this.path.size
    }
    fun isSubPath(path: ClaimsPointer) : Boolean {
        val currentPath = this.path.toMutableList()
        if (currentPath.isEmpty() && path.path.firstOrNull() != null) {
            return false
        }
        for(p in path.path) {
            val cp = currentPath.removeFirstOrNull()
            // path is longer than this, hence it is a subpath (as we did not early return before)
            if(cp == null) {
                return true
            }
            when(cp) {
                // Index only is equal if both are index and the index is the same
                is PointerPart.Index  if(p is PointerPart.Index && cp.v1 == p.v1) -> { continue  }
                // Null is only valid on arrays, which means path can only be a subpath if it is an index or null
                // which also means it CANNOT be PointerPart.String
                is PointerPart.Null if(p is PointerPart.Null || p is PointerPart.Index ) -> { continue }
                is PointerPart.String if(p is PointerPart.String && p.v1 == cp.v1) ->  { continue }
                else -> {
                    return false
                }
            }
        }
        return this.path.size == path.path.size
    }
    override fun select(v: Value): List<Value> {
        var currentlySelected = mutableListOf<Value>(v)
        for(part in path) {
            when(part) {
                is PointerPart.String if currentlySelected.all { it.isObject() } -> {
                    currentlySelected = currentlySelected.map { it.get(part) }.filter { it !is Value.Null}.toMutableList()
                }
                is PointerPart.Index if currentlySelected.all { it.isArrayLike() } -> {
                    currentlySelected = currentlySelected.map { it.get(part) }.filter { it !is Value.Null}.toMutableList()
                }
                is PointerPart.Null if currentlySelected.all { it.isArrayLike() } -> {
                    currentlySelected = currentlySelected.flatMap { it.getAll() }.filter { it !is Value.Null}.toMutableList()
                }
                else -> throw QueryException.InvalidType()
            }
        }
        return currentlySelected
    }

    override fun resolvePtr(v: Value): List<List<PointerPart>> {
        var currentPointers = mutableListOf<MutableList<PointerPart>>(mutableListOf())
        var thePointer = mutableListOf<PointerPart>()
        for (part in path) {
            if (part is PointerPart.Null) {
                val element = ClaimsPointer(thePointer).select(v)
                if (element.size > 1 || element.isEmpty()) {
                    return emptyList()
                }
                if(!element[0].isArrayLike()) {
                    return emptyList()
                }
                val elementSize = element[0].getAll().size
                val newPointers = mutableListOf<MutableList<PointerPart>>()
                for (ptrs in currentPointers) {
                    for(i in 0..<elementSize) {
                        val p = ptrs.toMutableList()
                        p.add(PointerPart.Index(i.toULong()))
                        newPointers.add(p)
                    }
                }
                currentPointers = newPointers
            } else {
                for(ptr in currentPointers) {
                    ptr.add(part)
                }
            }
            thePointer.add(part)
        }
        return currentPointers
    }

    override fun equals(other: Any?): Boolean {
        if(other !is ClaimsPointer) {
            return false
        }
        return this.path == other.path
    }
}

operator fun Value.get(part: PointerPart): Value {
    return when (part) {
        is PointerPart.Index -> this[part.v1.toInt()]
        is PointerPart.Null -> Value.Null
        is PointerPart.String -> this[part.v1]
    }
}
operator fun Value.get(selector: Selector): List<Value> {
    return selector.select(this)
}

fun List<PointerPart>.asSelector(): Selector {
    return ClaimsPointer(this)
}
fun String.asSelector(): Selector {
    return ClaimsPointer(listOf(PointerPart.String(this)))
}
fun Int.asSelector(): Selector {
    return ClaimsPointer(listOf(PointerPart.Index(this.toULong())))
}

fun ClaimsPointer.toPointer() : List<PointerPart> {
    return this.path
}

fun List<Any?>.toClaimsPointer() : ClaimsPointer? {
    var elements = mutableListOf<PointerPart>()
    for (e in this) {
        when (e) {
            null ->  {
                elements.add(PointerPart.Null(false))
                continue
            }
            is String -> {
                elements.add(PointerPart.String(e))
                continue
            }
            is Int -> {
                elements.add(PointerPart.Index(e.toULong()))
                continue
            }
            is PointerPart -> {
                elements.add(e)
                continue
            }
            else -> {
                return null
            }
        }
    }
    return ClaimsPointer(elements)
}

File: kapun-credential-core/rust/src/claims_pointer.rs
/* Copyright 2025 Ubique Innovation AG

Licensed under the Apache License, Version 2.0 (the "License");
you may not use this file except in compliance with the License.
You may obtain a copy of the License at

  http://www.apache.org/licenses/LICENSE-2.0

Unless required by applicable law or agreed to in writing,
software distributed under the License is distributed on an
"AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
KIND, either express or implied.  See the License for the
specific language governing permissions and limitations
under the License.
 */

use std::fmt::{Display, Formatter};

use crate::models::{Pointer, PointerPart};
use kapun_util_rust::value::Value;

#[derive(Debug, uniffi::Error)]
pub enum QueryError {
    InvalidType,
    InvalidIndex,
    NoElementsFound,
}
impl Display for QueryError {
    fn fmt(&self, f: &mut Formatter<'_>) -> std::fmt::Result {
        f.write_str("{self:?}")
    }
}
#[uniffi::export(with_foreign)]
pub trait Selector: Send + Sync {
    fn select(&self, v: Value) -> Result<Vec<Value>, QueryError>;
    fn resolve_ptr(&self, v: Value) -> Result<Vec<Pointer>, QueryError>;
}

impl Selector for Pointer {
    fn select(&self, v: Value) -> Result<Vec<Value>, QueryError> {
        let s = selector(self);
        s(&v)
    }

    fn resolve_ptr(&self, v: Value) -> Result<Vec<Pointer>, QueryError> {
        let mut current_pointers = vec![vec![]];
        let mut the_pointer = vec![];
        for p in self {
            match p {
                PointerPart::Null(_) => {
                    let element = the_pointer.select(v.clone())?;
                    if element.len() > 1 || element.is_empty() {
                        return Ok(vec![]);
                    }
                    if !element[0].is_array() {
                        return Ok(vec![]);
                    }
                    let element_size = element[0].as_array().unwrap().len();
                    let mut new_pointers = vec![];
                    for ptrs in &current_pointers {
                        for i in 0..element_size {
                            let mut p = ptrs.clone();
                            p.push(PointerPart::Index(i as u64));
                            new_pointers.push(p)
                        }
                    }
                    current_pointers = new_pointers;
                }
                _ => {
                    for ptr in &mut current_pointers {
                        ptr.push(p.clone())
                    }
                }
            }
            the_pointer.push(p.clone());
            let _ = the_pointer.select(v.clone())?;
        }
        Ok(current_pointers)
    }
}

pub fn selector(path: &Pointer) -> impl Fn(&Value) -> Result<Vec<Value>, QueryError> + '_ {
    move |input| {
        let mut currently_selected = vec![input.clone()];
        for part in path {
            match part {
                crate::models::PointerPart::String(key)
                    if currently_selected
                        .iter()
                        .all(|a| a.is_object() || a.is_ordered_object()) =>
                {
                    currently_selected = currently_selected
                        .iter()
                        .flat_map(|a| a.get(key))
                        .cloned()
                        .collect()
                }
                crate::models::PointerPart::Index(i)
                    if currently_selected.iter().all(|a| a.is_array()) =>
                {
                    currently_selected = currently_selected
                        .iter()
                        .flat_map(|a| a.get(*i as usize))
                        .cloned()
                        .collect()
                }
                crate::models::PointerPart::Null(_)
                    if currently_selected.iter().all(|a| a.is_array()) =>
                {
                    currently_selected = currently_selected
                        .iter()
                        .filter_map(|a| a.as_array())
                        .flatten()
                        .cloned()
                        .collect()
                }
                _ => return Err(QueryError::InvalidType),
            }
            if currently_selected.is_empty() {
                return Err(QueryError::NoElementsFound);
            }
        }
        Ok(currently_selected
            .into_iter()
            .filter(|a| !a.is_null())
            .collect())
    }
}

#[macro_export]
macro_rules! pointer {
    ($($e:expr),+) => {
        vec![$(
            $crate::models::PointerPart::from($e),
        )*
        ]
    };
}

#[cfg(test)]
mod tests {
    use crate::{claims_pointer::QueryError, models::PointerPart};

    use super::selector;

    #[test]
    pub fn test_empty() {
        let data = serde_json::json!({
            "test" : [{
                "name" : ["a", "c"]
            },{
                "name" : ["b", "d"]
            }]
        });
        let query0: Vec<PointerPart> =
            serde_json::from_str(r#"["test", null, "name", 4, "a"]"#).unwrap();
        let s1 = selector(&query0);
        let err = s1(&data.into()).unwrap_err();
        assert!(matches!(dbg!(err), QueryError::NoElementsFound))
    }
    #[test]
    pub fn test_invalid_type() {
        let data = serde_json::json!({
            "test" : [{
                "name" : ["a", "c"]
            },{
                "name" : ["b", "d"]
            }]
        });
        let query0: Vec<PointerPart> =
            serde_json::from_str(r#"["test", null, 0, 1, "a"]"#).unwrap();
        let s1 = selector(&query0);
        let err = s1(&data.into()).unwrap_err();
        assert!(matches!(dbg!(err), QueryError::InvalidType))
    }
    #[test]
    pub fn test_null_only_array() {
        let data = serde_json::json!({
            "test" : [{
                "name" : ["a", "c"]
            },{
                "name" : ["b", "d"]
            }]
        });
        let query0: Vec<PointerPart> =
            serde_json::from_str(r#"["test", null, null, 1, "a"]"#).unwrap();
        let s1 = selector(&query0);
        let err = s1(&data.into()).unwrap_err();
        assert!(matches!(dbg!(err), QueryError::InvalidType))
    }

    #[test]
    fn simple_test() {
        let data = serde_json::json!({
            "test" : [{
                "name" : ["a", "c"]
            },{
                "name" : ["b", "d"]
            }]
        });
        let pointer = vec![
            PointerPart::String("test".to_string()),
            PointerPart::Null(None),
            PointerPart::String("name".to_string()),
            PointerPart::Index(0),
        ];
        let s = selector(&pointer);
        let result = s(&data.into()).unwrap();
        let a = result[0].as_str().unwrap();
        let b = result[1].as_str().unwrap();
        assert_eq!(a, "a");
        assert_eq!(b, "b");
    }

    #[test]
    pub fn test_from_string() {
        let query: Vec<PointerPart> = serde_json::from_str(r#"["test", null, "name", 0]"#).unwrap();
        let pointer = vec![
            PointerPart::String("test".to_string()),
            PointerPart::Null(None),
            PointerPart::String("name".to_string()),
            PointerPart::Index(0),
        ];
        assert_eq!(query, pointer);
    }

    #[test]
    pub fn nested_arrays() {
        let query0: Vec<PointerPart> =
            serde_json::from_str(r#"["test", null, 0, null, "a"]"#).unwrap();
        let query1: Vec<PointerPart> =
            serde_json::from_str(r#"["test", null, 1, null, "a"]"#).unwrap();
        let data = serde_json::json!({
            "test" : [
                [
                    [
                        {"a": "test1"},
                        {"a": "test2"}
                    ],
                    [
                        {"a": "test3"},
                        {"a": "test4"}
                    ]
                ],[
                    [
                        {"a": "test1"},
                        {"a": "test2"}
                    ],
                    [
                        {"a": "test3"},
                        {"a": "test4"}
                    ]
                ]
            ]
        });

        let s1 = selector(&query0);
        let s2 = selector(&query1);

        let result1 = s1(&data.clone().into()).unwrap();
        let result2 = s2(&data.into()).unwrap();

        assert_eq!(result1[0].as_str().unwrap(), "test1");
        assert_eq!(result1[1].as_str().unwrap(), "test2");
        assert_eq!(result1[2].as_str().unwrap(), "test1");
        assert_eq!(result1[3].as_str().unwrap(), "test2");

        println!("{result2:?}");
        assert_eq!(result2[0].as_str().unwrap(), "test3");
        assert_eq!(result2[1].as_str().unwrap(), "test4");
        assert_eq!(result2[2].as_str().unwrap(), "test3");
        assert_eq!(result2[3].as_str().unwrap(), "test4");

        assert_eq!(result1.len(), 4);
        assert_eq!(result2.len(), 4);
    }
}
